File: ast/statement/if.go
package statement

import (
	"github.com/panda-io/micro-panda/ast"
	"github.com/panda-io/micro-panda/ast/core"
	"github.com/panda-io/micro-panda/ast/types"
)

type If struct {
	StatementBase
	Initialization core.Statement
	Condition      ast.Expression
	Body           core.Statement
	Else           core.Statement
}

func (i *If) Validate(c ast.Context) {
	ctx := c.NewContext()
	if i.Initialization != nil {
		i.Initialization.Validate(ctx)
	}
	if i.Condition == nil {
		c.Error(i.GetPosition(), "expect condition expression")
	} else {
		i.Condition.Validate(ctx, types.TypeBool)
		if i.Condition.Type() != nil && !i.Condition.Type().Equal(types.TypeBool) {
			c.Error(i.Condition.GetPosition(), "expect bool type condition")
		}
	}
	if i.Body != nil {
		bodyCtx := ctx.NewContext()
		i.Body.Validate(bodyCtx)
	}
	if i.Else != nil {
		elseCtx := ctx.NewContext()
		i.Else.Validate(elseCtx)
	}
}

File: MicroPanda/AST/Statement/If.cs
namespace MicroPanda.AST.Statement;

using Expression;
using Type;

internal class If : Statement
{
    internal Statement? Initialization { get; set; }
    internal Expression? Condition { get; set; }
    internal Statement? Body { get; set; }
    internal Statement? Else { get; set; }

    internal override void Validate(Context context)
    {
        var ctx = context.NewContext();
        Initialization?.Validate(ctx);
        if (Condition == null)
        {
            context.Program.Error(Position, "expect condition expression");
        }
        else
        {
            Condition.Validate(ctx, TypeHelper.TypeBool);
            if (Condition.Type != null && !Condition.Type.Equal(TypeHelper.TypeBool))
            {
                context.Program.Error(Condition.Position, "expect bool type condition");
            }
        }
        if (Body != null)
        {
            var bodyCtx = ctx.NewContext();
            Body.Validate(bodyCtx);
        }
        if (Else != null)
        {
            var elseCtx = ctx.NewContext();
            Else.Validate(elseCtx);
        }
    }
}